File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/qr/ZXingQRProcessor.java
package hu.kdea.szavazas.ballotprocessor.qr;

import boofcv.struct.image.GrayU8;
import com.google.zxing.BinaryBitmap;
import com.google.zxing.MultiFormatReader;
import com.google.zxing.NotFoundException;
import com.google.zxing.RGBLuminanceSource;
import com.google.zxing.Result;
import com.google.zxing.ResultPoint;
import com.google.zxing.common.GlobalHistogramBinarizer;
import com.google.zxing.common.HybridBinarizer;
import hu.kdea.szavazas.ballotprocessor.Logger;
import hu.kdea.szavazas.ballotprocessor.common.RectangleData;

import java.util.function.Consumer;
import javax.inject.Inject;
public class ZXingQRProcessor  {
    @Inject
    public ZXingQRProcessor() {
    }

    public void apply(GrayU8 image, Consumer<QrResult> onResult) {
        int width = image.width;
        int height = image.height;
        int[] pixels = new int[width * height];
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                int value = image.get(x, y);
                pixels[y * width + x] = (0xFF << 24) | (value << 16) | (value << 8) | value;
            }
        }
        Logger.INSTANCE.d(
            "ZXingQR",
            "Image size: " + width + "x" + height + ", first pixel: " + image.get(0, 0) + ", last pixel: " + image.get(width - 1, height - 1)
        );
        RGBLuminanceSource source = new RGBLuminanceSource(width, height, pixels);
        MultiFormatReader reader = new MultiFormatReader();
        Exception lastException = null;
        String[] names = {"Hybrid", "GlobalHistogram"};
        for (String name : names) {
            try {
                BinaryBitmap binaryBitmap = createBinaryBitmap(source, name);
                Result result = reader.decode(binaryBitmap);
                Logger.INSTANCE.d("ZXingQR", "Decoded with " + name + ": " + result.getText());
                String raw = result.getText();
                String[] parts = raw.split("-");
                int numSupport = parts.length > 1 ? parseOrDefault(parts[1], 3) : 3;
                int numCandidates = parts.length > 2 ? parseOrDefault(parts[2], 11) : 11;
                ResultPoint[] points = result.getResultPoints();
                int minX = Integer.MAX_VALUE;
                int minY = Integer.MAX_VALUE;
                int maxX = Integer.MIN_VALUE;
                int maxY = Integer.MIN_VALUE;
                for (ResultPoint point : points) {
                    int x = (int) point.getX();
                    int y = (int) point.getY();
                    minX = Math.min(minX, x);
                    minY = Math.min(minY, y);
                    maxX = Math.max(maxX, x);
                    maxY = Math.max(maxY, y);
                }
                RectangleData bbox = new RectangleData(
                    minX == Integer.MAX_VALUE ? 0 : minX,
                    minY == Integer.MAX_VALUE ? 0 : minY,
                    maxX == Integer.MIN_VALUE ? 1 : maxX - minX + 1,
                    maxY == Integer.MIN_VALUE ? 1 : maxY - minY + 1
                );
                onResult.accept(new QrResult(raw, numSupport, numCandidates, bbox));
                return;
            } catch (NotFoundException exception) {
                Logger.INSTANCE.d("ZXingQR", "QR not found with " + name);
                lastException = exception;
            } catch (Exception exception) {
                Logger.INSTANCE.e("ZXingQR", "Unexpected error with " + name + ": " + exception.getMessage());
                lastException = exception;
            }
        }
        Logger.INSTANCE.e("ZXingQR", "All binarizers failed. Last error: " + (lastException == null ? null : lastException.getMessage()));
        onResult.accept(null);
    }

    private BinaryBitmap createBinaryBitmap(RGBLuminanceSource source, String name) {
        if ("Hybrid".equals(name)) {
            return new BinaryBitmap(new HybridBinarizer(source));
        }
        return new BinaryBitmap(new GlobalHistogramBinarizer(source));
    }

    private int parseOrDefault(String value, int defaultValue) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException exception) {
            return defaultValue;
        }
    }
}

File: test-utils/src/main/java/hu/kdea/szavazas/ballotprocessor/ZXingQRProcessor.kt
package hu.kdea.szavazas.ballotprocessor

import boofcv.struct.image.GrayU8
import com.google.zxing.*
import com.google.zxing.common.HybridBinarizer
import com.google.zxing.common.GlobalHistogramBinarizer

class ZXingQRProcessor : IQRProcessor {

    override fun detect(image: GrayU8, onResult: (QrResult?) -> Unit) {
        val width = image.width
        val height = image.height
        val pixels = IntArray(width * height)
        for (y in 0 until height) {
            for (x in 0 until width) {
                val v = image.get(x, y)
                pixels[y * width + x] = (0xFF shl 24) or (v shl 16) or (v shl 8) or v
            }
        }

        Logger.d(
            "ZXingQR",
            "Image size: ${width}x${height}, first pixel: ${
                image.get(
                    0,
                    0
                )
            }, last pixel: ${image.get(width - 1, height - 1)}"
        )

        val source = RGBLuminanceSource(width, height, pixels)
        val reader = MultiFormatReader()
        var lastException: Exception? = null

        // Try two common binarizers
        for ((name, binarizerFactory) in listOf(
            "Hybrid" to { BinaryBitmap(HybridBinarizer(source)) },
            "GlobalHistogram" to { BinaryBitmap(GlobalHistogramBinarizer(source)) }
        )) {
            try {
                val binaryBitmap = binarizerFactory()
                val result = reader.decode(binaryBitmap)
                Logger.d("ZXingQR", "Decoded with $name: ${result.text}")

                val raw = result.text
                val parts = raw.split("-")
                val numSupport = parts.getOrNull(1)?.toIntOrNull() ?: 3
                val numCandidates = parts.getOrNull(2)?.toIntOrNull() ?: 11

                val points = result.resultPoints
                val xs = points.map { it.x.toInt() }
                val ys = points.map { it.y.toInt() }
                val bbox = Rect(
                    xs.minOrNull() ?: 0,
                    ys.minOrNull() ?: 0,
                    (xs.maxOrNull() ?: 0) - (xs.minOrNull() ?: 0) + 1,
                    (ys.maxOrNull() ?: 0) - (ys.minOrNull() ?: 0) + 1
                )

                onResult(QrResult(raw, numSupport, numCandidates, bbox))
                return
            } catch (e: NotFoundException) {
                Logger.d("ZXingQR", "QR not found with $name")
                lastException = e
            } catch (e: Exception) {
                Logger.e("ZXingQR", "Unexpected error with $name: ${e.message}")
                lastException = e
            }
        }

        Logger.e("ZXingQR", "All binarizers failed. Last error: ${lastException?.message}")
        onResult(null)
    }
}